cmd/collectors/restperf/plugins/volumetopclients/volumetopclients.go

```diff
@@ -0,0 +1,409 @@
+		return err
+	}
+
+	if maxVol := v.Params.GetChildContentS("MaxVolumeCount"); maxVol != "" {
+		if maxVolCount, err := strconv.Atoi(maxVol); err != nil {
+			v.maxVolumeCount = defaultTopN
+		} else {
+			v.maxVolumeCount = int(math.Min(float64(maxVolCount), float64(maxTopN)))
+		}
+	}
+	v.Logger.Info().Int("maxVolumeCount", v.maxVolumeCount).Msg("Using maxVolumeCount")
+	return nil
+}
+
+func (v *Volume) Run(dataMap map[string]*matrix.Matrix) ([]*matrix.Matrix, *util.Metadata, error) {
+	data := dataMap[v.Object]
+	v.client.Metadata.Reset()
+	err := v.InitAllMatrix()
+	if err != nil {
+		return nil, nil, err
+	}
+	for k := range v.data {
+		// Set all global labels if already not exist
+		v.data[k].SetGlobalLabels(data.GetGlobalLabels())
+	}
+	err = v.processTopClients(data)
+	if err != nil {
+		return nil, nil, err
+	}
+	result := make([]*matrix.Matrix, 0, len(v.data))
+
+	var pluginInstances uint64
+	for _, value := range v.data {
+		instCount := len(value.GetInstances())
+		if instCount == 0 {
+			continue
+		}
+		result = append(result, value)
+		pluginInstances = +uint64(len(value.GetInstances()))
+	}
+
+	v.client.Metadata.PluginInstances = pluginInstances
+	return result, v.client.Metadata, err
+}
+
+func (v *Volume) getCachedVolumesWithActivityTracking() (*set.Set, error) {
+	const cacheDuration = time.Hour
+
+	// Check if the cache is still valid
+	if v.cache != nil && time.Since(v.cache.lastFetched) < cacheDuration {
+		return v.cache.volumesWithActivityTrackingEnabled, nil
+	}
+
+	va, err := v.fetchVolumesWithActivityTrackingEnabled()
+	if err != nil {
+		return nil, err
+	}
+
+	// Update the cache
+	v.cache = &VolumeCache{
+		volumesWithActivityTrackingEnabled: va,
+		lastFetched:                        time.Now(),
+	}
+
+	return va, nil
+}
+
+func (v *Volume) processTopClients(data *matrix.Matrix) error {
+	va, err := v.getCachedVolumesWithActivityTracking()
+	if err != nil {
+		return err
+	}
+
+	if va.Size() == 0 {
+		return nil
+	}
+
+	filteredDataInstances := set.New()
+	for key, value := range data.GetInstances() {
+		svmName := value.GetLabel("svm")
+		volName := value.GetLabel("volume")
+		if va.Has(svmName + keyToken + volName) {
+			filteredDataInstances.Add(key)
+		}
+	}
+
+	readOpsList, writeOpsList, readDataList, writeDataList := v.collectMetricValues(data, filteredDataInstances)
+
+	topReadOps := v.getTopN(readOpsList)
+	topWriteOps := v.getTopN(writeOpsList)
+	topReadData := v.getTopN(readDataList)
+	topWriteData := v.getTopN(writeDataList)
+
+	readOpsVolumes, readOpsSvms := v.extractVolumesAndSvms(data, topReadOps)
+	writeOpsVolumes, writeOpsSvms := v.extractVolumesAndSvms(data, topWriteOps)
+	readDataVolumes, readDataSvms := v.extractVolumesAndSvms(data, topReadData)
+	writeDataVolumes, writeDataSvms := v.extractVolumesAndSvms(data, topWriteData)
+
+	if err := v.processTopClientsByMetric(readOpsVolumes, readOpsSvms, topClientReadOPSMatrix, "iops.read", v.setOpsMetric); err != nil {
```

Comment:
setOpsMetric and setDataMetric can be consolidated into a single function if you pass along "ops" or "data" to `processTopClientsByMetric`. That also means you don't need to pass the set function either